multitenant: reduce touch points required to add new capabilities

ecwall opened the pull request:
Fixes #97652

Currently a adding a new capability of an existing type requires modifying call
sites in many files:
```
1) pkg/multitenant/tenantcapabilities/tenantcapabilitiestestutils/testutils.go
2) pkg/sql/crdb_internal.go
3) pkg/sql/show_tenant.go
4) pkg/sql/tenant_capability.go
5) pkg/testutils/testcluster/testcluster.go
```

This change allows new capabilities of existings types to be added without
modifying these call sites - they only need to be modified if a new capability
type is added.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/97469)
<!-- Reviewable:end -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97469)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @knz, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go` line 86 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NP4COqABa-8ZjhxnCSK:-NP4COqABa-8ZjhxnCSL:b-xlogdz) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d747e97fef46847a36567d1413976156f7c7ab9/pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go#L86)):*
> ```Go
> 		}
> 		if !cp.GetFlagCapability(capabilityName) {
> 			return errors.Newf("tenant %s does not have capability %q", tenID, capabilityName)
> ```

This makes the error consistent for all cases and includes the exact capability name in the output `"can_admin_split"` instead of a variation `admin split`.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiespb/capabilities.go` line 68 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NP4ChY8Cy6ivUMskr2K:-NP4ChY8Cy6ivUMskr2L:b-xv8abe) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d747e97fef46847a36567d1413976156f7c7ab9/pkg/multitenant/tenantcapabilities/tenantcapabilitiespb/capabilities.go#L68)):*
> ```Go
> //go:generate stringer -type=TenantCapabilityName -linecomment
> const (
> 	// CanAdminSplit maps to TenantCapabilities.CanAdminSplit.
> ```

I changed these to reference instead of copy the source to prevent the comments from becoming desycned.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiespb/capabilities.go` line 76 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NP4Czhx4EAbAJ-mxWcg:-NP4Czhx4EAbAJ-mxWch:bi04faq) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d747e97fef46847a36567d1413976156f7c7ab9/pkg/multitenant/tenantcapabilities/tenantcapabilitiespb/capabilities.go#L76)):*
> ```Go
> )
> 
> func (t *TenantCapabilities) getFlagFieldRef(capabilityName TenantCapabilityName) *bool {
> ```

Only this switch and the AdminBatchRequest switch need to be modified when new capabilities are added.


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97469)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @knz, and @rafiss)

___
*[`pkg/sql/multitenant_admin_function_test.go` line 311 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NP4DL0sCjD_neucTl-J:-NP4DL0sCjD_neucTl-K:bqg3ea5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d747e97fef46847a36567d1413976156f7c7ab9/pkg/sql/multitenant_admin_function_test.go#L311)):*
> ```Go
> 	setCapabilities := func(
> 		tenantID roachpb.TenantID,
> 		capabilityNames ...tenantcapabilitiespb.TenantCapabilityName,
> ```

Changed this naming to be consistent with other usages of `tenantcapabilitiespb.TenantCapabilityName`.


<!-- Sent from Reviewable.io -->

ajwerner commented:
Who are the ideal reviewers for this? I'm working on adding a new capability, and it's not a boolean, and so I skew heavily with this.

ajwerner commented:
>  It looks like the PR you worked on didn't handle these yet unless you added that after I looked at it.

That PR doesn't. I'm working on the next one 😉

arulajmani reviewed (commented):
Would it be beneficial to reduce these touchpoints, and more generally the refactors this PR is proposing, once Andrew's PR lands? That way, we have a holistic, concrete view of how the subsystem is looking. Additionally, it also allows us to add capabilities in parallel without stepping on each other's toes. 

> Reviewed 1 of 7 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-:-NPsdDot5CO4I3UPFoaT:bxefjq6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @ecwall, @knz, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiesapi/api.go` line 11 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NPscG4R05KUDL9BTpAu:-NPscG4SDGZMfI_W09iZ:bsr7rf0) ([raw file](https://github.com/cockroachdb/cockroach/blob/8955c937e261d6a119e5aba93e0b4125fd430e23/pkg/multitenant/tenantcapabilities/tenantcapabilitiesapi/api.go#L11)):*
> ```Go
> // licenses/APL.txt.
> 
> package tenantcapabilitiesapi
> ```

Why this new package? I would expect the `TenantCapabilities` interface to be in the `tenantcapabilities` base package. 
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiesapi/api.go` line 80 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NPscPHWE-GNCMeMiE7o:-NPscPHWE-GNCMeMiE7p:b-fh81jt) ([raw file](https://github.com/cockroachdb/cockroach/blob/8955c937e261d6a119e5aba93e0b4125fd430e23/pkg/multitenant/tenantcapabilities/tenantcapabilitiesapi/api.go#L80)):*
> ```Go
> type TenantCapabilities interface {
> 	// GetBoolCapability returns the value of the corresponding bool capability.
> 	GetBoolCapability(BoolCapabilityName) bool
> ```

Copying over my suggestion from the internal slack thread, this should be capability focused instead of type focused. 
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiespb/capabilities.go` line 68 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NP4ChY8Cy6ivUMskr2K:-NPscawO0sZ-T2scrqDj:b-this1g) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d747e97fef46847a36567d1413976156f7c7ab9/pkg/multitenant/tenantcapabilities/tenantcapabilitiespb/capabilities.go#L68)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

I changed these to reference instead of copy the source to prevent the comments from becoming desycned.
</blockquote></details>

I don't follow your comment, could you say more?
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/testdata/basic` line 22 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NPscp031Yyu7J_n5avX:-NPscp031Yyu7J_n5avY:b-9js53s) ([raw file](https://github.com/cockroachdb/cockroach/blob/8955c937e261d6a119e5aba93e0b4125fd430e23/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/testdata/basic#L22)):*
> ```
> Incremental Update
> update: ten=10 cap=can_admin_split:true 
> update: ten=11 cap=
> ```

Did we break these tests?


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
I don't think that this PR should conflict much with @ajwerner's. I saw his was approved so I can rebase this on his if he is merging it soon.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-:-NPshkyH1NnwYt3KJdK8:b-r2cehm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, @knz, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiesapi/api.go` line 11 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NPscG4R05KUDL9BTpAu:-NPsgcBvEhgjZMyJ5fzA:b-ylsf6q) ([raw file](https://github.com/cockroachdb/cockroach/blob/8955c937e261d6a119e5aba93e0b4125fd430e23/pkg/multitenant/tenantcapabilities/tenantcapabilitiesapi/api.go#L11)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Why this new package? I would expect the `TenantCapabilities` interface to be in the `tenantcapabilities` base package. 
</blockquote></details>

It looks like the base package has unrelated dependencies like `pkg/kv/kvpb` because there pattern is to create a new authorizer function per call site where the call site passes in its own models instead of a general purpose API that any call site can use so the base package is likely to bring in more and more dependencies over time that capability users would bring in transitively.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiespb/capabilities.go` line 68 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NP4ChY8Cy6ivUMskr2K:-NPseZHZ4hTXfPEMWZ5s:bgrj5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d747e97fef46847a36567d1413976156f7c7ab9/pkg/multitenant/tenantcapabilities/tenantcapabilitiespb/capabilities.go#L68)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I don't follow your comment, could you say more?
</blockquote></details>

This line was different when I made the comment so it no longer applies here. The comment applies to the comments on the enum values referencing the protobuf fields.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/testdata/basic` line 22 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NPscp031Yyu7J_n5avX:-NPsesLn-sK8ZCEbTceE:b2rgzqm) ([raw file](https://github.com/cockroachdb/cockroach/blob/8955c937e261d6a119e5aba93e0b4125fd430e23/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/testdata/basic#L22)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Did we break these tests?
</blockquote></details>

The tests relied on a struct being formatted with `%+v`, but now the struct is an interface so the output of `String()` changed.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (dismissed):
> I don't think that this PR should conflict much with @ajwerner's. I saw his was approved so I can rebase this on his if he is merging it soon.

Sorry for not being clear, by Andrew's PR I meant the followup PR he referenced above -- the one which actually sets/shows `SpanConfigBounds`. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-:-NPt2gA77dfOwpDf4oMB:bnvg5vt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @ecwall, @knz, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiesapi/api.go` line 11 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NPscG4R05KUDL9BTpAu:-NPt5WQIBzfxBZ6H5BGQ:b-nqgtv4) ([raw file](https://github.com/cockroachdb/cockroach/blob/8955c937e261d6a119e5aba93e0b4125fd430e23/pkg/multitenant/tenantcapabilities/tenantcapabilitiesapi/api.go#L11)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

It looks like the base package has unrelated dependencies like `pkg/kv/kvpb` because there pattern is to create a new authorizer function per call site where the call site passes in its own models instead of a general purpose API that any call site can use so the base package is likely to bring in more and more dependencies over time that capability users would bring in transitively.
</blockquote></details>

Could you clarify the concern here? I don't follow at all -- is `pkg/kv/kvpb` causing some sort of dependency cycle? 
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/testdata/basic` line 22 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NPscp031Yyu7J_n5avX:-NPt4t_t5RgpoFpzGzg5:b5jgptt) ([raw file](https://github.com/cockroachdb/cockroach/blob/8955c937e261d6a119e5aba93e0b4125fd430e23/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/testdata/basic#L22)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

The tests relied on a struct being formatted with `%+v`, but now the struct is an interface so the output of `String()` changed.
</blockquote></details>

Are you happy with the new datadriven output as a result of your change? Specifically, this `cap=` stuff?


<!-- Sent from Reviewable.io -->

ecwall committed:
multitenant: reduce touch points required to add new capabilities

Fixes #97652

Currently a adding a new capability of an existing type requires modifying call
sites in many files:
```
1) pkg/multitenant/tenantcapabilities/tenantcapabilitiestestutils/testutils.go
2) pkg/sql/crdb_internal.go
3) pkg/sql/show_tenant.go
4) pkg/sql/tenant_capability.go
5) pkg/testutils/testcluster/testcluster.go
```

This change allows new capabilities of existings types to be added without
modifying these call sites - they only need to be modified if a new capability
type is added.

Release note: None

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97469)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, @knz, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/testdata/basic` line 22 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NPscp031Yyu7J_n5avX:-NPwuPdTASief9EtA8YA:b7gmb6f) ([raw file](https://github.com/cockroachdb/cockroach/blob/d675f83b8969cf993d8e8127fbdaea34a94c38cf/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/testdata/basic#L22)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Are you happy with the new datadriven output as a result of your change? Specifically, this `cap=` stuff?
</blockquote></details>

It looks like it is using the output of `proto.CompactTextString` now which is fine with me.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiesapi/api.go` line 11 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NPscG4R05KUDL9BTpAu:-NPwuoq8-ussT7Dhx6Fm:blhwn2h) ([raw file](https://github.com/cockroachdb/cockroach/blob/d675f83b8969cf993d8e8127fbdaea34a94c38cf/pkg/multitenant/tenantcapabilities/tenantcapabilitiesapi/api.go#L11)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Could you clarify the concern here? I don't follow at all -- is `pkg/kv/kvpb` causing some sort of dependency cycle? 
</blockquote></details>

I merged the `tenantcapabilitiesapi` package into `tenantcapabilities`.

This pattern increases the chances of a cycle happening because packages depending on the `tenantcapabilities` package end up getting some of each others dependencies transitively.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
All right now this PR works for me. There's a few questions here and there and I believe it will be good to go.

I also remember @arulajmani may have a suggestion to enhance the Get/Put API further. I'll be curious to see how that looks too.

> Reviewed 5 of 29 files at r3, 3 of 7 files at r4, 2 of 2 files at r5, 2 of 18 files at r6, 27 of 30 files at r7, 1 of 1 files at r8, 2 of 4 files at r9, 1 of 2 files at r10, 3 of 3 files at r11, 1 of 2 files at r12, 1 of 1 files at r13, 1 of 1 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-:-NPxOhuQDt-76qITeGUK:b-bi1z5d)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, @ecwall, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go` line 74 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NPxNR_Y7mA35yISAQWq:-NPxNR_Y7mA35yISAQWr:bk008u8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2eafac05b3308320fe1d61b8fd057fb8177dc606/pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go#L74)):*
> ```Go
> 		}
> 		if !hasCap || requiredCap == onlySystemTenant || !found || !cp.GetBool(requiredCap) {
> 			if requiredCap == tenantcapabilities.CanAdminSplit && a.knobs.AuthorizerSkipAdminSplitCapabilityChecks {
> ```

I thought we discussed this testing knob was intended for removal?
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go` line 134 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NPxNaRWBgHIfCYULAds:-NPxNaRWBgHIfCYULAdt:b-clre5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/2eafac05b3308320fe1d61b8fd057fb8177dc606/pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go#L134)):*
> ```Go
> 	reflect.TypeOf((*kvpb.MergeRequest)(nil)):                         onlySystemTenant,
> 	reflect.TypeOf((*kvpb.RequestLeaseRequest)(nil)):                  onlySystemTenant,
> 	reflect.TypeOf((*kvpb.LeaseInfoRequest)(nil)):                     noCapCheckNeeded,
> ```

If you are sure about this one and the one below, please move them to the group of requests above (and ensure the entries are still sorted).
___
*[`pkg/sql/tenant_capability.go` line 84 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NPxNuntEsEz0dwKNb_Q:-NPxNuntEsEz0dwKNb_R:b-l00lnd) ([raw file](https://github.com/cockroachdb/cockroach/blob/2eafac05b3308320fe1d61b8fd057fb8177dc606/pkg/sql/tenant_capability.go#L84)):*
> ```Go
> 			if capability.Value == nil {
> 				// TODO: Uncomment this block when a new capability type is added above.
> 				//  It is commented out to prevent a linter error.
> ```

What's up with this? what's the linter error?
___
*[`pkg/rpc/auth_test.go` line 252 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NPxOWmf1WbBYXhOJKIK:-NPxOWmf1WbBYXhOJKIL:bsw06i8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb2940708105dc742d8210a54bd3ddc099e62ad/pkg/rpc/auth_test.go#L252)):*
> ```Go
> 		return makeReqShared(t, key, endKey...)
> 	}
> 	makeDisallowedAdminReq := func(key string) kvpb.Request {
> ```

curious: why did you remove these test cases?


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97469)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, @knz, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go` line 74 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NPxNR_Y7mA35yISAQWq:-NPxT4Px82KqlMf0V9W3:bu0g17z) ([raw file](https://github.com/cockroachdb/cockroach/blob/2eafac05b3308320fe1d61b8fd057fb8177dc606/pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go#L74)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

I thought we discussed this testing knob was intended for removal?
</blockquote></details>

A bunch of files unrelated to the those in this PR need to be changed so I was going to do it separately.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go` line 134 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NPxNaRWBgHIfCYULAds:-NPxUVUX01cl8SHMPSlR:bftfmdx) ([raw file](https://github.com/cockroachdb/cockroach/blob/2eafac05b3308320fe1d61b8fd057fb8177dc606/pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go#L134)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

If you are sure about this one and the one below, please move them to the group of requests above (and ensure the entries are still sorted).
</blockquote></details>

Yeah these had to change to get existing tests passing. I will relocate them.
___
*[`pkg/sql/tenant_capability.go` line 84 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NPxNuntEsEz0dwKNb_Q:-NPxV0n7BviwEC7W9qaI:b3jalwm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2eafac05b3308320fe1d61b8fd057fb8177dc606/pkg/sql/tenant_capability.go#L84)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

What's up with this? what's the linter error?
</blockquote></details>

`missingValueDefault == nil` is always false until we add a type above that does not set `missingValueDefault`.
___
*[`pkg/rpc/auth_test.go` line 252 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NPxOWmf1WbBYXhOJKIK:-NPxVCQe20Yj42KFEIRY:b-2ftb7w) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb2940708105dc742d8210a54bd3ddc099e62ad/pkg/rpc/auth_test.go#L252)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

curious: why did you remove these test cases?
</blockquote></details>

They were testing the switch that was moved out of here and into the authorizer. In this test file, the equivalent behavior is now tested by setting `authorizer.hasCapabilityForBatch = false` in this test case:

```
			{
				req: &kvpb.BatchRequest{Requests: makeReqs(
					makeReqShared(t, prefix(10, "a"), prefix(10, "b")),
				)},
				configureAuthorizer: func(authorizer *mockAuthorizer) {
					authorizer.hasCapabilityForBatch = false
				},
				expErr: "tenant does not have capability",
			},
```


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
I feel satisfied with the PR now.
I also learned of Arul's additional request to use kvpb.Method for performance. I agree this is a reasonable request. I was simply not aware of that interface when we drafted the changes yesterday. My approval is conditional on that change. We can arrange aesthetic changes in followup PRs.

> Reviewed 1 of 1 files at r16, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-:-NPy4FMN80U_Sy_oNJi-:bei9y4v)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, @ecwall, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go` line 74 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NPxNR_Y7mA35yISAQWq:-NPy3pH2ApbJ82C5hJn1:b-o5o2de) ([raw file](https://github.com/cockroachdb/cockroach/blob/2eafac05b3308320fe1d61b8fd057fb8177dc606/pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go#L74)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

A bunch of files unrelated to the those in this PR need to be changed so I was going to do it separately.
</blockquote></details>

thanks. Is there a tracking issue?


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
> Reviewed 8 of 14 files at r18, 1 of 1 files at r19, 3 of 5 files at r20, 1 of 1 files at r21, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97469)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, @ecwall, and @rafiss)



<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97469)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, @knz, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go` line 74 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NPxNR_Y7mA35yISAQWq:-NQ0eC9C4RFcy4dlRjGw:bynalbg) ([raw file](https://github.com/cockroachdb/cockroach/blob/2eafac05b3308320fe1d61b8fd057fb8177dc606/pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go#L74)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

thanks. Is there a tracking issue?
</blockquote></details>

This is the issue, but it mentions a our initial plan to use a migration so I think it needs to be updated: https://github.com/cockroachdb/cockroach/issues/96736.

I have a followup PR (currently a draft that needs to be rebased) to this one to fallback to a default:  https://github.com/cockroachdb/cockroach/pull/97852


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
Like we talked about offline, I had 2 blocking changes here -- one was that we use kvpb.Method for perf reasons, and the other was that we continue to keep the datadriven tests in the tenantcapabilities package readable. As such, I'll remove the "blocking" portion of my review. 

Also like we talked about offline, I'm okay if we decide to merge the first commit and continue to iterate the interfaces here later (as new capabilities, with different types, are added here). I haven't given this a thorough review, and there's already an approval here, but let me know if you want me to give it a detailed look. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-:-NQ2RN2c4Fcd5cCFA92x:b9kfhq5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @ecwall, @knz, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 118 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NQ2SB9o54VfhMP5-hp8:-NQ2SB9o54VfhMP5-hp9:b-g7wgj8) ([raw file](https://github.com/cockroachdb/cockroach/blob/5207de82fd2a7528903ca08164d3b3af92eebb73/pkg/multitenant/tenantcapabilities/capabilities.go#L118)):*
> ```Go
> }
> 
> // CapabilityID represents a handle to a tenant capability.
> ```

nit: consider renaming this to `CapabilityName` instead. 
___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 119 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NQ2S3ba3AKDvZhvDZB5:-NQ2S3ba3AKDvZhvDZB6:b-u91iaa) ([raw file](https://github.com/cockroachdb/cockroach/blob/5207de82fd2a7528903ca08164d3b3af92eebb73/pkg/multitenant/tenantcapabilities/capabilities.go#L119)):*
> ```Go
> 
> // CapabilityID represents a handle to a tenant capability.
> type CapabilityID uint8
> ```

nit: Should this be uint to allow for more than 8 capabilities? 
___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 148 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NQ2QwXaCqVwL2u9wkKh:-NQ2QwXaCqVwL2u9wkKi:b6db9yr) ([raw file](https://github.com/cockroachdb/cockroach/blob/5207de82fd2a7528903ca08164d3b3af92eebb73/pkg/multitenant/tenantcapabilities/capabilities.go#L148)):*
> ```Go
> // TenantCapabilities is the interface provided by the capability store,
> // to provide access to capability values.
> type TenantCapabilities interface {
> ```

After our discussion, did you want to get rid of this interface before checking it in? Or did you want to keep it around?

If we get rid of it, does it mean all that changes is `Capability` looks something like:

```
// Capability is the accessor to a capability's current value.
type Capability interface {
	// Get retrieves the current value of the capability.
	Get(tenantcapabilitiespb.TenantCapabilities) Value
	// Set modifies the current value of the capability. Note that
	// calling Set does not persist the result. Persistence needs to be
	// arranged separately by storing the underlying protobuf; this is
	// done using a different interface. See the code for ALTER TENANT.
	Set(tenantcapabilitiespb.TenantCapabilities, interface{})
}

```

I have no strong feelings either way, just pointing it out. 


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97469)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, @knz, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 118 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NQ2SB9o54VfhMP5-hp8:-NQ2VsGs173IjWcY0eGU:b-lzrpye) ([raw file](https://github.com/cockroachdb/cockroach/blob/5207de82fd2a7528903ca08164d3b3af92eebb73/pkg/multitenant/tenantcapabilities/capabilities.go#L118)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: consider renaming this to `CapabilityName` instead. 
</blockquote></details>

@knz chose this name which I think is more self documenting because it identifies a capability for a tenant meaning there cannot be duplicates, but I do not have a strong opinion either way. @knz do you have thoughts?
___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 119 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NQ2S3ba3AKDvZhvDZB5:-NQ2WPus2bMF3cvMUJvv:bobyg7r) ([raw file](https://github.com/cockroachdb/cockroach/blob/5207de82fd2a7528903ca08164d3b3af92eebb73/pkg/multitenant/tenantcapabilities/capabilities.go#L119)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: Should this be uint to allow for more than 8 capabilities? 
</blockquote></details>

A `uint8` capability can have 2^8 - 1 (0 is reserved) values
___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 148 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NQ2QwXaCqVwL2u9wkKh:-NQ2X24JDidI3j1-7uq7:blfx833) ([raw file](https://github.com/cockroachdb/cockroach/blob/5207de82fd2a7528903ca08164d3b3af92eebb73/pkg/multitenant/tenantcapabilities/capabilities.go#L148)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

After our discussion, did you want to get rid of this interface before checking it in? Or did you want to keep it around?

If we get rid of it, does it mean all that changes is `Capability` looks something like:

```
// Capability is the accessor to a capability's current value.
type Capability interface {
	// Get retrieves the current value of the capability.
	Get(tenantcapabilitiespb.TenantCapabilities) Value
	// Set modifies the current value of the capability. Note that
	// calling Set does not persist the result. Persistence needs to be
	// arranged separately by storing the underlying protobuf; this is
	// done using a different interface. See the code for ALTER TENANT.
	Set(tenantcapabilitiespb.TenantCapabilities, interface{})
}

```

I have no strong feelings either way, just pointing it out. 
</blockquote></details>

I will leave it for now, but we can change it in the next PR.

I think it makes sense to have 2 getters
1) `Value ` wrapper for string usages - this is the simplest to use because you don't need to handle types differently
2) Raw value for custom usages - more complex, but gives you access to the actual protobuf data
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/testdata/basic` line 22 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NPscp031Yyu7J_n5avX:-NQ2Xyaf2GrCBjKCsRWa:b-bduy92) ([raw file](https://github.com/cockroachdb/cockroach/blob/d675f83b8969cf993d8e8127fbdaea34a94c38cf/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/testdata/basic#L22)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

It looks like it is using the output of `proto.CompactTextString` now which is fine with me.
</blockquote></details>

Updated to include zero-valued fields.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97469)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @ecwall, @knz, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 119 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NQ2S3ba3AKDvZhvDZB5:-NQ2p-3qCVGg6ynxddR3:b-1z398h) ([raw file](https://github.com/cockroachdb/cockroach/blob/5207de82fd2a7528903ca08164d3b3af92eebb73/pkg/multitenant/tenantcapabilities/capabilities.go#L119)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

A `uint8` capability can have 2^8 - 1 (0 is reserved) values
</blockquote></details>

:man_facepalming: 


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97469)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, @knz, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 118 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/97469#-NQ2SB9o54VfhMP5-hp8:-NQ2uOlO58j37pnTqNNz:b-keirqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/5207de82fd2a7528903ca08164d3b3af92eebb73/pkg/multitenant/tenantcapabilities/capabilities.go#L118)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

@knz chose this name which I think is more self documenting because it identifies a capability for a tenant meaning there cannot be duplicates, but I do not have a strong opinion either way. @knz do you have thoughts?
</blockquote></details>

I created this to address it separately so this can be merged https://github.com/cockroachdb/cockroach/issues/98273


<!-- Sent from Reviewable.io -->

ecwall commented:
bors r=knz
